test/e2e/external_gateways.go

```diff
@@ -481,5 +644,99 @@ func setupGatewayContainers(f *framework.Framework, nodes *v1.NodeList, gwContai
 			setupListenersOrDie(containerName, addressesv6.targetIPs[0])
 		}
 	}
+
+	for _, containerName := range []string{gwContainer1, gwContainer2} {
+		for _, postCreation := range postCreations {
+			postCreation(containerName)
+		}
+	}
 	return addressesv4, addressesv6
 }
+
+func annotatePodForGateway(podName, namespace, networkIPs string, bfd bool) {
+	// add the annotations to the pod to enable the gateway forwarding.
+	// this fakes out the multus annotation so that the pod IP is
+	// actually an IP of an external container for testing purposes
+	annotateArgs := []string{
+		"annotate",
+		"pods",
+		podName,
+		fmt.Sprintf("k8s.v1.cni.cncf.io/network-status=[{\"name\":\"%s\",\"interface\":"+
+			"\"net1\",\"ips\":[%s],\"mac\":\"%s\"}]", "foo", networkIPs, "01:23:45:67:89:10"),
+		fmt.Sprintf("k8s.ovn.org/routing-namespaces=%s", namespace),
+		fmt.Sprintf("k8s.ovn.org/routing-network=%s", "foo"),
+	}
+	if bfd {
+		annotateArgs = append(annotateArgs, "k8s.ovn.org/bfd-enabled=\"\"")
+	}
+	framework.Logf("Annotating the external gateway pod with annotation %s", annotateArgs)
+	framework.RunKubectlOrDie("default", annotateArgs...)
+}
+
+func hostNamesForContainers(containers []string) map[string]struct{} {
+	res := make(map[string]struct{})
+	for _, c := range containers {
+		hostName := hostNameForContainer(c)
+		res[hostName] = struct{}{}
+	}
+	return res
+}
+
+func hostNameForContainer(container string) string {
+	res, err := runCommand("docker", "exec", container, "hostname")
+	framework.ExpectNoError(err, "failed to run hostname on", container)
+	framework.Logf("Hostname for %s is %s", container, res)
+	return strings.TrimSuffix(res, "\n")
+}
+
+func pokeHostnameViaNC(podName, namespace, protocol, target string, port int) string {
+	args := []string{"exec", podName, "--"}
+	if protocol == "tcp" {
+		args = append(args, "bash", "-c", fmt.Sprintf("echo | nc -w 1 %s %d", target, port))
```

Comment:
why do you need to pipe in echo here?

Reply:
To get something reach the other end, so it replies with the hostname. This is how all the "nc based" tests currently work.